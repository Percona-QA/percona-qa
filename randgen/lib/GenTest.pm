package GenTest;
use base 'Exporter';
@EXPORT = ('say', 'tmpdir', 'safe_exit', 'windows', 'xml_timestamp', 'rqg_debug');

use strict;

use Cwd;
use POSIX;

my $tmpdir;

1;

sub BEGIN {
	foreach my $tmp ($ENV{TMP}, $ENV{TEMP}, $ENV{TMPDIR}, '/tmp', '/var/tmp', cwd()."/tmp" ) {
		if (
			(defined $tmp) &&
			(-e $tmp)
		) {
			$tmpdir = $tmp;
			last;
		}
	}

	if (
		($^O eq 'MSWin32') ||
		($^O eq 'MSWin64')
	) {
		$tmpdir = $tmpdir.'\\';
	} else {
		$tmpdir = $tmpdir.'/';
	}

	die("Unable to locate suitable temporary directory.") if not defined $tmpdir;
	
	return 1;
}

sub new {
	my $class = shift;
	my $args = shift;

	my $obj = bless ([], $class);

        my $max_arg = (scalar(@_) / 2) - 1;

        foreach my $i (0..$max_arg) {
                if (exists $args->{$_[$i * 2]}) {
                        $obj->[$args->{$_[$i * 2]}] = $_[$i * 2 + 1];
                } else {
                        warn("Unkown argument '$_[$i * 2]' to ".$class.'->new()');
                }
        }

        return $obj;
}

sub say {
	my @t = localtime();
	my $text = shift;

	if ($text =~ m{[\r\n]}sio) {
	        foreach my $line (split (m{[\r\n]}, $text)) {
			print "# ".sprintf("%02d:%02d:%02d", $t[2], $t[1], $t[0])." $line\n";
		}
	} else {
		print "# ".sprintf("%02d:%02d:%02d", $t[2], $t[1], $t[0])." $text\n";
	}
}

sub tmpdir {
	return $tmpdir;
}

sub safe_exit {
	my $exit_status = shift;
	POSIX::_exit($exit_status);
}

sub windows {
	if (
		($^O eq 'MSWin32') ||
	        ($^O eq 'MSWin64')
	) {
		return 1;
	} else {
		return 0;
	}	
}

sub xml_timestamp {
	my $datetime = shift;

	my ($sec, $min, $hour, $mday, $mon, $year, $wday, $yday, $isdst) = defined $datetime ? localtime($datetime) : localtime();
	$mday++;
	$year += 1900;
	
	return sprintf("%04d-%02d-%02dT%02d:%02d:%02dZ", $year, $mon ,$mday ,$hour, $min, $sec);
	
}

sub rqg_debug {
	if ($ENV{RQG_DEBUG}) {
		return 1;
	} else {
		return 0;
	}
}

1;
