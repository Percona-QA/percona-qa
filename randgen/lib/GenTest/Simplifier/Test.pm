package GenTest::Simplifier::Test;

require Exporter;
use GenTest;
@ISA = qw(GenTest);

use strict;

use lib 'lib';
use GenTest::Simplifier::Tables;

use constant SIMPLIFIER_EXECUTORS	=> 0;
use constant SIMPLIFIER_QUERIES		=> 1;
use constant SIMPLIFIER_RESULTS		=> 2;

my @optimizer_variables = (
	'optimizer_switch',
	'optimizer_use_mrr',
	'engine_condition_pushdown',
	'join_cache_level'
);

1;

sub new {
        my $class = shift;

	my $simplifier = $class->SUPER::new({
		executors	=> SIMPLIFIER_EXECUTORS,
		results		=> SIMPLIFIER_RESULTS,
		queries		=> SIMPLIFIER_QUERIES
	}, @_);

	return $simplifier;
}

sub simplify {
	my $simplifier = shift;

	my $test;

	my $executors = $simplifier->executors();
	my $results = $simplifier->results();
	my $queries = $simplifier->queries();

	# If we have two Executors determine the differences in Optimizer settings and print them as test comments
	# If there is only one executor, dump its settings directly into the test as test queries

	if (defined $executors->[1]) {
		my $version1 = $executors->[0]->version();
		my $version2 = $executors->[1]->version();

		if ($version1 ne $version2) {
			$test .= "# Server0: version = $version1\n";
			$test .= "# Server1: version = $version2\n\n";
		}

		foreach my $optimizer_variable (@optimizer_variables) {
			my @optimizer_values;
			foreach my $i (0..1) {
				my $optimizer_value = $executors->[$i]->dbh()->selectrow_array('SELECT @@'.$optimizer_variable);
				$optimizer_value = 'ON' if $optimizer_value == 1 && $optimizer_variable eq 'engine_condition_pushdown';
				$optimizer_values[$i] = $optimizer_value;
			}

			if ($optimizer_values[0] ne $optimizer_values[1]) {
				$test .= "# The value of $optimizer_variable is distinct between the two servers:\n";
				foreach my $i (0..1) {
					if ($optimizer_values[$i] =~ m{^\d+$}) {
						$test .= "# Server $i : SET SESSION $optimizer_variable = $optimizer_values[$i];\n";
					} else {
						$test .= "# Server $i : SET SESSION $optimizer_variable = '$optimizer_values[$i]';\n";
					}
				}
			} else {
				$test .= "# The value of $optimizer_variable is common between the two servers:\n";
				$test .= "/*!50400 SET SESSION $optimizer_variable = $optimizer_values[0] */;\n";
			}

			$test .= "\n";
		}
		$test .= "\n\n";
	} elsif (defined $executors->[0]) {
		foreach my $optimizer_variable (@optimizer_variables) {
			my $optimizer_value = $executors->[0]->dbh->selectrow_array('SELECT @@'.$optimizer_variable);
			$optimizer_value = 'ON' if $optimizer_value == 1 && $optimizer_variable eq 'engine_condition_pushdown';

			if ($optimizer_value =~ m{^\d+$}) {
				$test .= "/*!50400 SET SESSION $optimizer_variable = $optimizer_value */;\n";
			} else {
				$test .= "/*!50400 SET SESSION $optimizer_variable = '$optimizer_value' */;\n";
			}
		}
		$test .= "\n\n";
	}

	my $query_count = defined $queries ? $#$queries : $#$results;

	foreach my $query_id (0..$query_count) {

		my $query;
		if (defined $queries) {
			$query = $queries->[$query_id];
		} else {
			$query = $results->[$query_id]->[0]->query();
		}

		$test .= "# Begin test case for query $query_id\n\n";

		my $simplified_database = 'query'.$query_id;

		my $tables_simplifier = GenTest::Simplifier::Tables->new(
			dsn		=> $executors->[0]->dsn(),
			orig_database	=> 'test',
			new_database	=> $simplified_database
		);

		my @participating_tables = $tables_simplifier->simplify($query);
		
		if ($#participating_tables > -1) {
			$test .= "--disable_warnings\n";
			$test .= "DROP TABLE IF EXISTS ".join(', ', @participating_tables).";\n";
			$test .= "--enable_warnings\n\n"
		}
			
		my $mysqldump_cmd = "mysqldump -uroot --no-set-names --compact --force --protocol=tcp --port=19306 $simplified_database ";
		$mysqldump_cmd .= join(' ', @participating_tables) if $#participating_tables > -1;
		open (MYSQLDUMP, "$mysqldump_cmd|");
		while (<MYSQLDUMP>) {
			next if $_=~ m{SET \@saved_cs_client}sio;
			next if $_=~ m{SET character_set_client}sio;
			$test .= $_;
		}
		close (MYSQLDUMP);

		$test .= "\n\n";

		$query =~ s{(SELECT|FROM|WHERE|GROUP\s+BY|ORDER\s+BY|HAVING|LIMIT)}{\n$1}sgio;
		$test .= $query.";\n\n";

		if (
			(defined $results) &&
			(defined $results->[$query_id])
		) {
			$test .= "# Diff:\n\n";

			# Add comments to each line in the diff, since MTR has issues with /* */ comment blocks.

			my $diff = GenTest::Comparator::dumpDiff(
				$simplifier->results()->[$query_id]->[0],
				$simplifier->results()->[$query_id]->[1]
			);

			$test .= "# ".join("\n# ", split("\n", $diff))."\n\n\n";
		}

		$test .= "DROP TABLE ".join(', ', @participating_tables).";\n\n" if $#participating_tables > -1;
	
		$test .= "# End of test case for query $query_id\n\n";
	}

	return $test;
}

sub executors {
	return $_[0]->[SIMPLIFIER_EXECUTORS];
}

sub queries {
	return $_[0]->[SIMPLIFIER_QUERIES];
}

sub results {
	return $_[0]->[SIMPLIFIER_RESULTS];
}

1;
