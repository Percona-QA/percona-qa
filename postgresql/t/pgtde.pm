package PGTDE;

use PostgreSQL::Test::Cluster;
use PostgreSQL::Test::Utils;

use File::Basename;
use File::Compare;
use Test::More;
use Time::HiRes qw(usleep);

# Expected .out filename of TAP testcase being executed. These are already part of repo under t/expected/*.
our $expected_filename_with_path;

# Result .out filename of TAP testcase being executed. Where needed, a new *.out will be created for each TAP test.
our $out_filename_with_path;

# Runtime output file that is used only for debugging purposes for comparison to PGSS, blocks and timings.
our $debug_out_filename_with_path;

my $expected_folder = "t/expected";
my $results_folder = "t/results";

sub psql
{
	my ($node, $dbname, $sql) = @_;

	my (undef, $stdout, $stderr) = $node->psql($dbname, $sql,
		extra_params => [ '-a', '-Pformat=aligned', '-Ptuples_only=off' ]);

	if ($stdout ne '')
	{
		append_to_result_file($stdout);
	}

	if ($stderr ne '')
	{
		append_to_result_file($stderr);
	}
}

# Copied from src/test/recovery/t/017_shm.pl
sub poll_start
{
	my ($node) = @_;

	my $max_attempts = 10 * $PostgreSQL::Test::Utils::timeout_default;
	my $attempts = 0;

	while ($attempts < $max_attempts)
	{
		$node->start(fail_ok => 1) && return 1;

		# Wait 0.1 second before retrying.
		usleep(100_000);

		# Clean up in case the start attempt just timed out or some such.
		$node->stop('fast', fail_ok => 1);

		$attempts++;
	}

	# Try one last time without fail_ok, which will BAIL_OUT unless it
	# succeeds.
	$node->start && return 1;
	return 0;
}

sub append_to_result_file
{
	my ($str) = @_;

	append_to_file($out_filename_with_path, $str . "\n");
}

sub append_to_debug_file
{
	my ($str) = @_;

	append_to_file($debug_out_filename_with_path, $str . "\n");
}

sub setup_files_dir
{
	my ($test_filename) = @_;

	unless (-d $results_folder)
	{
		mkdir $results_folder
		  or die "Can't create folder $results_folder: $!\n";
	}

	my ($test_name) = $test_filename =~ /([^.]*)/;

	$expected_filename_with_path = "${expected_folder}/${test_name}.out";
	$out_filename_with_path = "${results_folder}/${test_name}.out";
	$debug_out_filename_with_path =
	  "${results_folder}/${test_name}.out.debug";

	if (-f $out_filename_with_path)
	{
		unlink($out_filename_with_path)
		  or die
		  "Can't delete already existing $out_filename_with_path: $!\n";
	}
}

sub compare_results
{
	return compare($expected_filename_with_path, $out_filename_with_path);
}

sub backup
{
	my ($node, $backup_name, %params) = @_;
	my $backup_dir = $node->backup_dir . '/' . $backup_name;

	mkdir $backup_dir or die "mkdir($backup_dir) failed: $!";

	my $pg_tde_dir = $node->data_dir . '/pg_tde';
	if (-d $pg_tde_dir) {
		PostgreSQL::Test::RecursiveCopy::copypath($pg_tde_dir, $backup_dir . '/pg_tde');
	}
	else {
		note "Skipping pg_tde directory backup ?~@~T not present in data directory";
	}

	$node->backup($backup_name, %params);
}

sub setup_pg_tde_node {
    my ($node, $test_name) = @_;

    # Default test_name from script name if not provided
    $test_name ||= basename($0, '.pl');
    # Add process ID to ensure parallel safety
    my $pid = $$;
    # Build unique keyring file paths in /tmp
    my $global_keyring = File::Spec->catfile('/tmp', "global_keyring_${test_name}_${pid}.file");
    my $local_keyring  = File::Spec->catfile('/tmp', "local_keyring_${test_name}_${pid}.file");

    # Basic pg_tde settings
    $node->append_conf('postgresql.conf',
        "shared_preload_libraries = 'pg_tde'");
    $node->append_conf('postgresql.conf',
        "default_table_access_method = 'tde_heap'");

    $node->start;

    # Remove any existing keyring files for this test
    unlink($global_keyring_file);
    unlink($local_keyring_file);

    # Create and enable pg_tde extension
    $node->safe_psql('postgres',
        'CREATE EXTENSION IF NOT EXISTS pg_tde;');

	# Create global key provider and set server key
    $node->safe_psql('postgres',
        "SELECT pg_tde_add_global_key_provider_file(
            'global_key_provider', '$global_keyring_file');");

    $node->safe_psql('postgres',
        "SELECT pg_tde_create_key_using_global_key_provider(
            'global_test_key_time', 'global_key_provider');");

    $node->safe_psql('postgres',
        "SELECT pg_tde_set_server_key_using_global_key_provider(
            'global_test_key_time', 'global_key_provider');");

	# Create local key provider and set database key
    $node->safe_psql('postgres',
        "SELECT pg_tde_add_database_key_provider_file(
            'local_key_provider', '$local_keyring_file');");

    $node->safe_psql('postgres',
        "SELECT pg_tde_create_key_using_database_key_provider(
            'local_test_key_time', 'local_key_provider');");

    $node->safe_psql('postgres',
        "SELECT pg_tde_set_key_using_database_key_provider(
            'local_test_key_time', 'local_key_provider');");

    # WAL encryption setting
    my $WAL_ENCRYPTION = $ENV{WAL_ENCRYPTION} // 'on';
    $node->append_conf(
        'postgresql.conf',
        ($WAL_ENCRYPTION eq 'off')
            ? "pg_tde.wal_encrypt = off\n"
            : "pg_tde.wal_encrypt = on\n"
    );
}

1;
