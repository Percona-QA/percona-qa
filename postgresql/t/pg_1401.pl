#!/usr/bin/perl

use strict;
use warnings;
use File::Basename;
use Test::More;
use lib 't';
use pgtde;

PGTDE::setup_files_dir(basename($0));

unlink('/tmp/pg_tde_test_pg1401.per');
my $node = PostgreSQL::Test::Cluster->new('main');
$node->init;
$node->append_conf('postgresql.conf', "shared_preload_libraries = 'pg_tde'");
$node->start;

PGTDE::psql($node, 'postgres', 'CREATE EXTENSION IF NOT EXISTS pg_tde;');

PGTDE::psql($node, 'postgres',
	"SELECT pg_tde_add_global_key_provider_file('file-keyring-pg-1401','/tmp/pg_tde_test_pg1401.per');"
);
PGTDE::psql($node, 'postgres',
	"SELECT pg_tde_create_key_using_global_key_provider('server-key', 'file-keyring-pg-1401');"
);
PGTDE::psql($node, 'postgres',
	"SELECT pg_tde_set_key_using_global_key_provider('server-key', 'file-keyring-pg-1401');"
);

PGTDE::psql($node, 'postgres',
	'CREATE TABLE t1 (id SERIAL PRIMARY KEY,name VARCHAR(100),t2_id INT) using tde_heap;'
);

PGTDE::psql($node, 'postgres',
	"INSERT INTO t1(name) VALUES ('John'),('Mark');"
);

#Query the table
PGTDE::psql($node, 'postgres',
	"SELECT * FROM t1;"
);

#Change table access method to heap
PGTDE::psql($node, 'postgres',
	"ALTER TABLE t1 SET ACCESS METHOD heap;"
);

#Query the table
PGTDE::psql($node, 'postgres',
	"SELECT * FROM t1;"
);

PGTDE::append_to_result_file("-- Update postgresql.conf, remove pg_tde from shared_preload_libraries");
$node->adjust_conf('postgresql.conf', "shared_preload_libraries', ''");

PGTDE::append_to_result_file("-- server restart");
$node->restart;

#Query the table
PGTDE::psql($node, 'postgres',
	"SELECT * FROM t1;"
);

PGTDE::psql($node, 'postgres', 'DROP TABLE t1;');
PGTDE::psql($node, 'postgres', 'DROP EXTENSION pg_tde;');

$node->stop;

# Compare the expected and out file
my $compare = PGTDE->compare_results();

is($compare, 0,
	"Compare Files: $PGTDE::expected_filename_with_path and $PGTDE::out_filename_with_path files."
);

done_testing();
