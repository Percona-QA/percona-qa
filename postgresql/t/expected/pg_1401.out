CREATE EXTENSION IF NOT EXISTS pg_tde;
SELECT pg_tde_add_global_key_provider_file('file-keyring-pg-1401','/tmp/pg_tde_test_pg1401.per');
 pg_tde_add_global_key_provider_file 
-------------------------------------
                                  -1
(1 row)

SELECT pg_tde_set_key_using_global_key_provider('server-key', 'file-keyring-pg-1401');
 pg_tde_set_key_using_global_key_provider 
------------------------------------------
 
(1 row)

CREATE TABLE t1 (id SERIAL PRIMARY KEY,name VARCHAR(100),t2_id INT) using tde_heap;
INSERT INTO t1(name) VALUES ('John'),('Mark');
SELECT * FROM t1;
 id | name | t2_id 
----+------+-------
  1 | John |      
  2 | Mark |      
(2 rows)

ALTER TABLE t1 SET ACCESS METHOD heap;
SELECT * FROM t1;
 id | name | t2_id 
----+------+-------
  1 | John |      
  2 | Mark |      
(2 rows)

-- Update postgresql.conf, remove pg_tde from shared_preload_libraries
-- server restart
SELECT * FROM t1;
 id | name | t2_id 
----+------+-------
  1 | John |      
  2 | Mark |      
(2 rows)

DROP TABLE t1;
DROP EXTENSION pg_tde;
